File: deploy/agent_engine/agent.py
"""
TechParts Orchestrator — Vertex AI Agent Engine (Python)

Python equivalent of agents/orchestrator/src/agent.ts.
Connects to the three TypeScript workers on Cloud Run via A2A protocol.

A2A is language-agnostic: this Python orchestrator and the TypeScript workers
communicate over standard HTTP — neither side knows or cares what language
the other uses. This is the key cross-language multi-agent story.

Bug workaround:
  The TypeScript toA2a() server puts "http://localhost:8080/jsonrpc" in the
  agent card url because it doesn't know its own public Cloud Run URL at startup.
  When RemoteA2aAgent reads that card from Vertex AI Agent Engine, it tries to
  call localhost — which doesn't exist in the cloud.
  Fix: fetch the real card, patch the url to the correct public URL, pass it in.
  Long-term fix for the developer: add PUBLIC_URL env var support to shared/src/server.ts.
"""

import json
import os
import urllib.request

from a2a.types import AgentCard
from google.adk.agents import LlmAgent
from google.adk.agents.remote_a2a_agent import RemoteA2aAgent
from google.adk.tools.agent_tool import AgentTool

INVENTORY_URL = os.environ.get("INVENTORY_AGENT_URL", "")
ORDERS_URL = os.environ.get("ORDERS_AGENT_URL", "")
PRICING_URL = os.environ.get("PRICING_AGENT_URL", "")


def fetch_patched_card(base_url: str) -> AgentCard:
    """
    Fetch the A2A agent card from a Cloud Run worker and patch the url field.
    The TypeScript server always writes localhost into the card; we replace it
    with the real public Cloud Run URL so Vertex AI Agent Engine can reach it.
    """
    raw = json.loads(
        urllib.request.urlopen(f"{base_url}/.well-known/agent-card.json").read()
    )
    raw["url"] = f"{base_url}/jsonrpc"
    if "additionalInterfaces" in raw:
        for iface in raw["additionalInterfaces"]:
            transport = iface.get("transport", "")
            if "JSONRPC" in transport:
                iface["url"] = f"{base_url}/jsonrpc"
            elif "HTTP" in transport:
                iface["url"] = f"{base_url}/rest"
    return AgentCard(**raw)


inventory_agent = RemoteA2aAgent(
    name="inventory_agent",
    description="Knows the TechParts product catalog: products, prices, stock levels and warehouse locations. Ask it to find products or check stock.",
    agent_card=fetch_patched_card(INVENTORY_URL),
)

orders_agent = RemoteA2aAgent(
    name="orders_agent",
    description="Knows TechParts customer orders: order history, order details, and return eligibility under the 30-day policy. Ask it about customers and orders.",
    agent_card=fetch_patched_card(ORDERS_URL),
)

pricing_agent = RemoteA2aAgent(
    name="pricing_agent",
    description="Compares TechParts prices with the market: our price for a product plus live competitor prices from the web.",
    agent_card=fetch_patched_card(PRICING_URL),
)

root_agent = LlmAgent(
    name="ops_orchestrator",
    model="gemini-2.5-flash",
    description="TechParts operations assistant for support staff. Coordinates the inventory, orders and pricing agents to resolve customer cases end-to-end.",
    instruction="""You are the TechParts operations assistant used by internal support staff.

You have no data of your own. Three specialist agents do the work; call them as tools and pass them clear, self-contained natural-language requests:
- orders_agent: customer order history, order details, return eligibility (30-day policy).
- inventory_agent: product catalog, prices, stock, alternatives.
- pricing_agent: whether our price for a product is competitive vs the market.

How to work a case:
1. Break the request into sub-questions and send each to the right specialist. Include all context the specialist needs (ids, SKUs, product names) — they do not see this conversation.
2. Use earlier answers to inform later calls (e.g. first find what the customer bought, then ask inventory for in-stock alternatives to that product).
3. Synthesize one clear recommendation for the support employee: what to tell the customer and what actions to take.

Rules:
- Never invent data; everything must come from the specialists.
- If a specialist reports a blocker (e.g. return window expired), say so and propose the best alternative.
- Answer as a short action plan with the key facts (order, eligibility, suggested replacement with stock/price, price competitiveness).""",
    tools=[
        AgentTool(agent=orders_agent),
        AgentTool(agent=inventory_agent),
        AgentTool(agent=pricing_agent),
    ],
)

File: agents/orchestrator/src/agent.ts
import { AgentTool, LoadMemoryTool, LlmAgent, RemoteA2AAgent } from '@google/adk';

const INVENTORY_URL = process.env.INVENTORY_AGENT_URL ?? 'http://localhost:8001';
const ORDERS_URL = process.env.ORDERS_AGENT_URL ?? 'http://localhost:8002';
const PRICING_URL = process.env.PRICING_AGENT_URL ?? 'http://localhost:8003';

// Each worker runs as its own A2A service. A RemoteA2AAgent connects to one over
// the network by resolving its agent card from the base URL
// (…/.well-known/agent-card.json). We wrap each in an AgentTool so the
// orchestrator can call it like any other tool — the difference from a local
// sub-agent is that the worker runs in a *separate process with its own session*
// and sees NONE of this conversation. Whatever a worker needs to know, we must
// pass in the call.
const inventoryAgent = new RemoteA2AAgent({
  name: 'inventory_agent',
  description:
    'Looks up TechParts products and live stock levels. Use for "do we carry X", "is X in stock", "how many units of X".',
  agentCard: INVENTORY_URL,
});

const ordersAgent = new RemoteA2AAgent({
  name: 'orders_agent',
  description:
    'Answers questions about specific customer orders and return eligibility. Use for order status, order history, and "can this be returned".',
  agentCard: ORDERS_URL,
});

const pricingAgent = new RemoteA2AAgent({
  name: 'pricing_agent',
  description:
    'Compares a TechParts product price against the current market using real web search. Use for "are we competitive on X", "what do competitors charge for X".',
  agentCard: PRICING_URL,
});

export const rootAgent = new LlmAgent({
  name: 'ops_orchestrator',
  model: process.env.GEMINI_MODEL ?? 'gemini-2.5-flash',
  description:
    'Coordinates the TechParts inventory, orders and pricing specialists to answer multi-part operational questions.',
  instruction: `You are the operations orchestrator for TechParts, a consumer-electronics retailer. You have no data of your own — you answer by delegating to specialist agents and synthesizing their replies.

Your specialists (each is a separate service that does NOT see this conversation):
- inventory_agent — products and live stock levels.
- orders_agent — specific customer orders and return eligibility.
- pricing_agent — competitiveness of our price vs. the live market (real web search).
- load_memory — recall relevant facts from earlier conversations with this user.

How to work, every time:
1. RECALL: call load_memory first with the key entities in the request (product, SKU, customer, order id). If a recent fact already answers part of the question (e.g. a competitor price we found before), prefer it over re-asking a specialist, but say it came from memory.
2. DECOMPOSE: break the request into the sub-questions each specialist can answer.
3. DELEGATE with FULL CONTEXT: call one specialist per sub-question. Because the specialist sees none of this conversation, every call must be self-contained — restate the product/SKU/order id/customer and any fact an earlier specialist gave you that this one needs. Never assume the specialist remembers anything.
4. SYNTHESIZE: combine the replies into one clear recommendation for internal staff. Cite which specialist each fact came from. If a specialist could not answer, say so rather than inventing the answer.

Be concise. Never invent products, prices, stock, orders or outcomes — every concrete fact must come from a specialist or from memory.`,
  tools: [
    new AgentTool({ agent: inventoryAgent }),
    new AgentTool({ agent: ordersAgent }),
    new AgentTool({ agent: pricingAgent }),
    new LoadMemoryTool(),
  ],
});
